test/ModuleInterface/package_interface.swift

```diff
@@ -0,0 +1,282 @@
+
+/// Client should not load a package interface module without -experimental-package-interface-load
+// RUN: not %target-swift-frontend -typecheck %t/Client.swift -I %t \
+// RUN:   -package-name barpkg \
+// RUN:   -Rmodule-loading 2> %t/load-pkg-off.output
+// RUN: %FileCheck -check-prefix=CHECK-LOAD-PKG-OFF %s < %t/load-pkg-off.output
+// CHECK-LOAD-PKG-OFF: loaded module 'Bar'; source: '{{.*}}/Bar.private.swiftinterface', loaded: '{{.*}}/Bar-{{.*}}.swiftmodule'
+// CHECK-LOAD-PKG-OFF: error: module 'Bar' is in package 'barpkg' but was built from a non-package interface; modules of the same package can only be loaded if built from source or package interface: {{.*}}/Bar.private.swiftinterface
+
+/// Client loads a private interface since the package-name is different from the loaded module's.
+// RUN: %target-swift-frontend -typecheck %t/Client.swift -I %t \
+// RUN:   -package-name foopkg \
+// RUN:   -experimental-package-interface-load \
+// RUN:   -Rmodule-loading 2> %t/load-diff-pkg.output
+// RUN: %FileCheck -check-prefix=CHECK-LOAD-DIFF-PKG %s < %t/load-diff-pkg.output
+// CHECK-LOAD-DIFF-PKG: loaded module 'Bar'; source: '{{.*}}/Bar.private.swiftinterface', loaded: '{{.*}}/Bar-{{.*}}.swiftmodule'
+
+// RUN: rm -rf %t/*.swiftmodule
+// RUN: rm -rf %t/Bar.package.swiftinterface
+
+/// Client loads a private interface since package interface doesn't exist. It should error since the loaded module is not built from a package interface.
+// RUN: not %target-swift-frontend -typecheck %t/Client.swift -I %t \
+// RUN:   -package-name barpkg \
+// RUN:   -experimental-package-interface-load \
+// RUN:   -Rmodule-loading 2> %t/load-priv.output
+// RUN: %FileCheck -check-prefix=CHECK-LOAD-PRIV %s < %t/load-priv.output
+// CHECK-LOAD-PRIV: loaded module 'Bar'; source: '{{.*}}/Bar.private.swiftinterface', loaded: '{{.*}}/Bar-{{.*}}.swiftmodule'
+// CHECK-LOAD-PRIV: error: module 'Bar' is in package 'barpkg' but was built from a non-package interface; modules of the same package can only be loaded if built from source or package interface: {{.*}}/Bar.private.swiftinterface
+
+// RUN: rm -rf %t/*.swiftmodule
+// RUN: rm -rf %t/Bar.private.swiftinterface
+
+/// Client loads a public interface since package or private interface doesn't exist.
+/// It should error since the loaded module is not built from a package interface.
+// RUN: not %target-swift-frontend -typecheck %t/Client.swift -I %t \
+// RUN:   -package-name barpkg \
+// RUN:   -experimental-package-interface-load \
+// RUN:   -Rmodule-loading 2> %t/load-pub.output
+
+// RUN: %FileCheck -check-prefix=CHECK-LOAD-PUB %s < %t/load-pub.output
+// CHECK-LOAD-PUB: loaded module 'Bar'; source: '{{.*}}/Bar.swiftinterface', loaded: '{{.*}}/Bar-{{.*}}.swiftmodule'
+// CHECK-LOAD-PUB: error: module 'Bar' is in package 'barpkg' but was built from a non-package interface; modules of the same package can only be loaded if built from source or package interface: {{.*}}/Bar.swiftinterface
```

Comment:
Could we make the test of the diagnostic a `-verify` check? It will ensure the error is on the import as expected.

Reply:
I'll update it in the followup PR that contains env var etc.